ImageInLib/ImageInLib/src/imageInterpolation.c

```diff
@@ -3,257 +3,145 @@
-                            newImage[kn][x] = originalImage[k][x];
-                        }
-                        else {
-                            newImage[kn][x] = originalImage[k + 1][x];
-                        }
-                    }
-                }
-            }
-            k_int = k_int + newSpacing;
-            kn = kn + 1;
-        } while (k_int < k2);
-    }
+Point3D getImageCoordFromRealCoord3D(Point3D image_point, Point3D image_origin, VoxelSpacing image_spacing, OrientationMatrix orientation) {
+    Point3D resultPoint;
+    resultPoint.x = image_origin.x + image_spacing.sx * orientation.v1.x * image_point.x;
+    resultPoint.y = image_origin.y + image_spacing.sy * orientation.v2.y * image_point.y;
+    resultPoint.z = image_origin.z + image_spacing.sz * orientation.v3.z * image_point.z;
+    return resultPoint;
+}
 
-    return true;
+Point3D getRealCoordFomImageCoord3D(Point3D real_point, Point3D real_origin, VoxelSpacing real_spacing, OrientationMatrix orientation) {
+    Point3D resultPoint;
+    resultPoint.x = (real_point.x - real_origin.x) / real_spacing.sx * orientation.v1.x;
+    resultPoint.y = (real_point.y - real_origin.y) / real_spacing.sy * orientation.v2.y;
+    resultPoint.z = (real_point.z - real_origin.z) / real_spacing.sz * orientation.v3.z;
+    return resultPoint;
 }
 
-bool linear2dInterpolation(dataType** originalImage, dataType** newImage, size_t imageLength, size_t imageWidth, size_t imageHeight, dataType originalSpacing, dataType newSpacing)
-{
-    if (originalImage == NULL || newImage == NULL)
-        return false;
+//=======================================================
+//2D images
 
-    size_t i, j, k, x, kn;
-    dataType k_int, k1, k2;
-    dataType divisionByOriginalSpacing = 1.0 / originalSpacing;
-
-    k_int = 0; kn = 0;
-    for (k = 0; k < imageHeight - 1; k++) {
-        k1 = k * originalSpacing;
-        k2 = k1 + originalSpacing;
-        do {
-            for (i = 0; i < imageLength; i++) {
-                for (j = 0; j < imageWidth; j++) {
-                    x = x_new(i, j, imageLength);
-                    newImage[kn][x] = (dataType)(originalImage[k][x] * ((k2 - k_int) * divisionByOriginalSpacing) +
-                        originalImage[k + 1][x] * ((k_int - k1) * divisionByOriginalSpacing));
-                }
-            }
-            k_int = k_int + newSpacing;
-            kn = kn + 1;
-        } while (k_int < k2);
-    }
+Point2D getRealCoordFromImageCoord2D(Point2D image_point, Point2D real_origin, PixelSpacing real_spacing, OrientationMatrix2D orientation) {
+    Point2D result_point;
+    result_point.x = real_origin.x + real_spacing.sx * orientation.v1.x * image_point.x;
+    result_point.y = real_origin.y + real_spacing.sy * orientation.v2.y * image_point.y;
+    return result_point;
+}
 
-    return true;
+Point2D getImageCoordFromRealCoord2D(Point2D real_point, Point2D real_origin, PixelSpacing real_spacing, OrientationMatrix2D orientation) {
```

Comment:
In my opinion, the formulas are not correct.
I think, they should be like this 
result_point.x = (real_point.x - real_origin.x) / (real_spacing.sx * orientation.v1.x);